gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java

```diff
@@ -200,13 +210,11 @@ private FileSystem getOptionallyThrottledFileSystem(FileSystem fs, State state)
 
       try {
 
-        Collection<? extends CopyEntity> files = this.copyableDataset.getCopyableFiles(this.targetFs, this.copyConfiguration);
-        List<FileSet<CopyEntity>> fileSets = partitionCopyableFiles(this.copyableDataset, files);
-
-        // Sort to optimize the insertion to work units list
-        Collections.sort(fileSets, this.workUnitList.getComparator());
+        Iterator<FileSet<CopyEntity>> fileSets =
+            this.copyableDataset.getFileSetIterator(this.targetFs, this.copyConfiguration);
 
-        for (FileSet<CopyEntity> fileSet : fileSets) {
+        while (fileSets.hasNext() && !this.workUnitList.hasRejectedFileSet()) {
```

Comment:
* Does `this.workUnitList.hasRejectedFileSet()` return true once a fileset is either rejected or evicted, or only if a fileset is rejected? If former is true, probably should rename to `hasRejectedOrEvictedFileSet`. If latter is true, why differentiating rejected vs evicted?
* If `this.workUnitList.hasRejectedFileSet()` returns false here, but becomes true before or while you create the workunit, is this a problem?